src/collectors/jolokia/kafka_jolokia.py

```diff
@@ -0,0 +1,592 @@
+
+        except Exception as e:
+            self.log.exception('Failed to collect consumer lag metrics from Zookeeper; ensure that your Kafka version is set correctly in the config file. These metrics will be SKIPPED, but processing will continue.')
+
+        ###
+        # 
+        # Now that we have collected the consumer group information from Zookeeper,
+        # the next steps are to collect the Zookeeper server stats and all of the
+        # Kafka performance metrics.
+        #
+        # For the Zookeeper server stats, we would like to collect via JMX for 
+        # consistency, but 2 of the metrics are not published to JMX. Therefore, 
+        # we will invoke the ZookeeperCollector, which talks directly to the
+        # Zookeeper port.
+        #
+        # For the Kakfa stats, well, that's the main purpose of this class, so we
+        # invoke the super-class collect() function on Jolokia to kick off the JMX 
+        # collection.
+        #
+        ###
+
+        try:
+            # The Zookeeper collector expects a parameter called "hosts" to list the hosts
+            # and port numbers to collect from.  Update the default config with this setting,
+            # based on the value we've already determined from our "zookeeper" parameter.
+            # Note that ="zookeeper@" is added to the front as the "alias", which will
+            # ensure that the metrics get listed under "kakfa.zookeeper."
+            self.config.update({'hosts': ['zookeeper@' + zookeeper]})
+
+            self.log.debug('Attempting to collect Zookeeper server metrics from hosts: ' + str(self.config.get('hosts')))
+
+            # Explicitly run the Zookeeper collector
+            ZookeeperCollector.collect(self)
+
+        except Exception as e:
+            self.log.exception('Failed to collect Zookeeper server metrics. These metrics will be SKIPPED, but processing will continue.')
+
+        try:
+            self.log.debug('Attempting to collect Kafka metrics via JMX.')
+
+            # And now run the Jolokia collector to get the JMX metrics
+            super(KafkaJolokiaCollector, self).collect()
+
+        except Exception as e:
+            self.log.exception('Failed to collect Kafka metrics via the Jolokia JMX bridge. These metrics will be SKIPPED, but processing will continue.')
+
+
+    ########
+    #
+    # The collect_consumer_lag_9() function collects the consumer lag metrics for Kafka
+    # versions 9 and above. Version 9 introduced the kafka.admin.ConsumerGroupCommand
+    # class which allows us to discover the consumer groups and the topics they are 
+    # using, rather than requiring them to be specified in the config file up front.
+    #
+    # 2016-11-03 DVG
+    #
+    ########
+
+    def collect_consumer_lag_9(self, zookeeper):
```

Comment:
It could be a single function with an if statement - if Kafka 8 do this else do that. I thought it was easier to read broken out into separate functions, but I can consolidate it if you prefer.  It won't make the code any shorter, though, as I've already factored out all the common code between the two into the process_lag_results() function.